rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,365 @@
+specification of `task_group`, accessing a `task_handle` after it is submitted
+via one of the run functions is undefined behavior. Currently, therefore, a 
+`task_handle` can only represent a created task. And currently, any task
+that is run can immediately be scheduled for execution since there is no notion
+of task dependencies for task_group.
+
+The first extension is to expand the semantics and usable lifetime of
+`task_handle` so that remains valid after it is passed to run and it can 
+represent tasks in any state, including submitted, executing, and completed 
+tasks. Similarly, a `task_handle` in the submitted state may represent a task
+that has predecessors that must complete before it can execute, and so passing
+a `task_handle` to `task_group::run` or `task_group::run_and_wait` only makes
+it available for dependency tracking, and does not make it immediately legal to
+execute.
+
+### Add function(s) to set dependencies.
+
+The obvious next extension is to add a mechanism for specifying dependencies
+between tasks. In the most conservative view, it should only be legal to add 
+additional predecessors / in-dependencies to tasks in the created state. 
+After a task starts is completed, it doesn’t make sense to add additional 
+predecessors, since it’s too late for them to delay the start of the task’s 
+execution. 
+
+It can make sense to add additional predecessors to a task that is
+currently executing if the executing task is suspended until those 
+additional dependencies complete. However, in this proposal we do not intend 
+to support this suspension model. 
+
+For a task in the submitted state, there can be a race between
+adding a new predecessor and the scheduler deciding to execute the task when its 
+currently known predecessors have completed. We will revisit the discussion of 
+adding predecessors to submitted tasks in the next section when we discuss 
+recursively grown task graphs. 
+
+Having mostly settled the question about when a predecessors can be added,
+then next question is what can be added as a predecessor task? The most 
+user-friendly answer is to have no limitation; any valid `task_handle` can act as 
+a predecessor. In many cases, a developer may only know what work must be completed 
+before a task can start but does not know the state of that work.
+
+We therefore think predecessors may be in any state when they are added, 
+as shown below:
+
+<img src="add_dependency.png" width=400>
+
+There are a number of possible options for the spelling of a function for adding 
+a single predecessor. We may also want a function to allow adding multiple 
+predecessors in one call.
+
+Given two `task_handle` objects `h1` and `h2`, some possible options 
+for adding `h1` as an in-dependence / predecessor of `h2` include:
+
+- `h2.add_predecessor(h1)`
+- `h2 = defer([]() { … }, h1)`
+- `make_edge(h1, h2)`
+
+We propose including the first option. Similarly, there could be
```

Comment:
I would prefer to not add methods to `task_handle` but use "external" functions, perhaps in the `task_group` class. This would be more consistent with the current approach (`defer`, `run`, run_and_wait`) as well as with `transfer_successors_to`.